### ISSUE TITLE
Agent does not retry if starting a stopped container fails

### ISSUE THREAD

**Issue by SpComb** [kontena/kontena#3293]
If a container already exists, is up-to-date and does not have any `.State.Error` set, but is stopped, then the agent `ServicePodWorker` will simply start it. This most commonly happens after a node reboot, but it can also happen when using `kontena service stop/start`. However, if the `#ensure_started` => `ServicePods::Starter` => `Docker::Container#start!` fails, then the agent will not attempt to re-start the container. The container will remain stopped until the service is re-deployed, or the agent is restarted. A `kontena service start` will not start the container either.

This happens in Kontena 1.2 - 1.4 because the `apply if needs_apply?` => `ensure_desired_state` relies on the `@container_state_changed` getting set to true if the container is not in the desired state, and only getting reset to `false` after a successful apply... it turns out that both of these assumptions are violated in this specific scenario of starting a container after a node reboot. This should already be fixed in Kontena 1.5 per #3259 replacing the `@container_state_changed` with `@apply_started/finished_at`, cutting out the Docker events from the feedback loop, allowing the `ServicePodWorker` to retry on any `apply` failures, not just those that lead to container events.

While the [`@container_state_changed` gets initialized to `true`](https://github.com/kontena/kontena/blob/v1.4.3/agent/lib/kontena/workers/service_pod_worker.rb#L32), this doesn't mean that the failing initial `#ensure_running` would leave it set to true... the [`ServicePodManager` does an initial `apply`](https://github.com/kontena/kontena/blob/v1.4.3/agent/lib/kontena/workers/service_pod_manager.rb#L118) at startup while the service pod is still in an `unknown` state, so the `ServicePodWorker@container_state_changed` will seemingly get reset to `false` on the `Kontena::Workers::ServicePodWorker: desired state is unknown for ..., not doing anything` 😒 The second `apply` once the `ServicePodManager` gets the real `service_pod.deploy_rev` from the master will then run with `@container_state_changed` set to false and remaining as false even if the `ensure_desired_state` fails.

Secondly, while a failed `#ensure_running` => `ServicePods::Creator` will trigger events like `create` for the new container, and thus set `@container_state_changed` to allow retrying per `needs_apply?`, a failed `Docker::Container#start!` will not trigger any container events via Docker, and the `@container_state_changed` will thus remain set to false.

**Comment by SpComb** [comment]
Repro case for testing this, using the Docker logging driver to cause the container start to fail if the target address is not resolvable:

```yaml
stack: test/log-syslog
services:
  foo:
    image: redis
  bar:
    image: redis
    logging:
      driver: syslog
      options:
        syslog-address: "udp://foo.${STACK}.${GRID}.kontena.local"
```

```
$ kontena stack install stacks/log-syslog.yaml 
 [done] Creating stack log-syslog      
 [done] Triggering deployment of stack log-syslog     
 [done] Waiting for deployment to start     
 [done] Deploying service foo     
 [done] Deploying service bar     
$ kontena service stop log-syslog/bar
 [done] Sending stop signal to log-syslog/bar service      
$ kontena service stop log-syslog/foo
 [done] Sending stop signal to log-syslog/foo service      
$ kontena service start log-syslog/bar
 [done] Sending start signal to log-syslog/bar service      
$ kontena service show log-syslog/bar
e2e/log-syslog/bar:
  created: 2018-02-19T11:38:58.559Z
  updated: 2018-02-19T11:38:58.559Z
  stack: e2e/log-syslog
  desired_state: running
  image: redis:latest
  revision: 1
  stack_revision: 1
  stateful: no
  scaling: 1
  strategy: ha
  read_only: no
  stop_grace_period: 10s
  deploy_opts:
  dns: bar.log-syslog.e2e.kontena.local
  log_driver: syslog
  log_opts:
    syslog-address: udp://foo.log-syslog.e2e.kontena.local
  instances:
    log-syslog/bar/1:
      scheduled_to: terom-e2e-dev
      deploy_rev: 2018-02-19 11:39:01 UTC
      rev: 2018-02-19 11:39:01 UTC
      state: missing
      error: Docker::Error::ServerError: failed to initialize logging driver: dial udp: lookup foo.log-syslog.e2e.kontena.local: no such host
      containers:
        log-syslog.bar-1 (on terom-e2e-dev):
          dns: bar-1.log-syslog.e2e.kontena.local
          ip: 10.81.128.51
          public ip: 18.195.116.41
          status: stopped
```

The service will succesfully start on the `install` because the `syslog-address` does exist at that point. After the `kontena service stop`, the service container will be stopped without any `.State.Error` being set. Trying to start it without the `foo.log-syslog.e2e.kontena.local
` service being running will cause the container start to fail, and then the container will remain stopped until deployed.

**Comment by SpComb** [comment]
Milestoning to 1.5 and closing so that this can get mentioned as a "Fixed issue" in the 1.5 release notes.
